git_machete/tests/functional/test_machete.py

```diff
@@ -25,6 +28,170 @@ def get_head_commit_hash() -> str:
     return os.popen("git rev-parse HEAD").read().strip()
 
 
+class FakeCommandLineOptions:
+    opt_debug: bool = False
+    opt_verbose: bool = False
+
+    def __init__(self) -> None:
+        self.opt_as_root: bool = False
+        self.opt_branch: Optional[str] = None
+        self.opt_checked_out_since: Optional[str] = None
+        self.opt_color: str = "auto"
+        self.opt_down_fork_point: Optional[str] = None
+        self.opt_draft: bool = False
+        self.opt_fetch: bool = False
+        self.opt_fork_point: Optional[str] = None
+        self.opt_inferred: bool = False
+        self.opt_list_commits: bool = False
+        self.opt_list_commits_with_hashes: bool = False
+        self.opt_merge: bool = False
+        self.opt_n: bool = False
+        self.opt_no_detect_squash_merges: bool = False
+        self.opt_no_edit_merge: bool = False
+        self.opt_no_interactive_rebase: bool = True
+        self.opt_onto: Optional[str] = None
+        self.opt_override_to: Optional[str] = None
+        self.opt_override_to_inferred: bool = False
+        self.opt_override_to_parent: bool = False
+        self.opt_push_tracked: Optional[bool] = True
+        self.opt_push_untracked: Optional[bool] = True
+        self.opt_return_to: str = "stay"
+        self.opt_roots: List[str] = list()
+        self.opt_start_from: str = "here"
+        self.opt_stat: bool = False
+        self.opt_sync_github_prs: bool = False
+        self.opt_unset_override: bool = False
+        self.opt_yes: bool = True
+
+
+class GitAPIState:
+    pulls: List[Dict[str, Any]] = [{'head': {'ref': 'bugfix/remove'}, 'user': {'login': 'github_user'}, 'base': {'ref': 'develop'}, 'number': '15', 'html_url': 'www.github.com'}]
+    user: Dict[str, str] = {'login': 'github_user', 'type': 'User', 'company': 'VirtusLab'}
+    issues: List[Dict[str, Any]] = []
+    remote_branches: List[str] = []
+
+    def __init__(self, url: str, headers: Dict[str, str] = None, data: Union[str, bytes, None] = None, method: str = ''):
```

Comment:
Looks nice <3

Reply:
If you have these variables as class variables aren't you modifying them in each tests and these modification persist from test to test?

This is rather risky, because now environment in which you're running your tests is dependent on what other tests have been run before.

Reply:
![uuu](https://user-images.githubusercontent.com/3383210/132480250-670ebcb6-2fb4-46d2-a23b-4fa18955e187.jpg)

Reply:
That was intended. New instance of `GithubAPIState` is created on each call to github API and each test makes several calls. It's not desired to lose this data. 
When needed, initial values of class variables are set on the beginning of test.

Reply:
I stand by my opinion that it's risky. I had in the past several tests in other project that were failing due to the sequence of the tests (they were not cleaning properly after themselves) and debugging/fixing such issues is a lengthy nightmare. If certain environment is expected before test or tests require certain setup you can put make a method out of it and start a test with it. If it's not required for test than we can clear it in teardown via `self.issues=[]`

Also if you require one test to be run after another one, it may break eventually while running with `stestr` or when one of the tests will be deleted/modified. It's asking for trouble in maintenance.

Reply:
I don't have time today to analyse it more deeply... but if that's the case that passing or failure of any test **depends on the order in which the tests are executed**, then I'm also strongly against such approach ☹️